geopandas/tools/overlay.py

```diff
@@ -198,12 +201,22 @@ def overlay(df1, df2, how="intersection", keep_geom_type=True):
             )
 
     # Computations
-    df1 = df1.copy()
-    df2 = df2.copy()
-    if df1.geom_type.isin(polys).all():
-        df1[df1._geometry_column_name] = df1.geometry.buffer(0)
-    if df2.geom_type.isin(polys).all():
-        df2[df2._geometry_column_name] = df2.geometry.buffer(0)
+    def preprocess(df):
```

Comment:
```suggestion
    def _make_valid(df):
    """ensure that geometries are valid using buffer(0)"""
```

preprocess is a bit general, better to use more explicit name for better readability.